csharp/ql/lib/semmle/code/csharp/Type.qll

```diff
@@ -45,6 +45,23 @@ class Type extends DotNet::Type, Member, TypeContainer, @type {
 
   /** Holds if this type is a value type, or a type parameter that is a value type. */
   predicate isValueType() { none() }
+
+  /** Gets an alternative type that is reported by the compiler as being ambiguous with this type. */
+  Type getAnAmbiguousAlternativeType() {
```

Comment:
Modeling it in this way means that we cannot distinguish when a type is used unambiguously and when it is used ambiguously, that may be too simplistic.

Reply:
Wouldn't `not exists(this.getAnAmbiguousAlternativeType())` mean that the type is used unambiguously?

Reply:
I may have misunderstood how it works, but consider this example:
```csharp
namespace N1 {
    class C {}
}

namespace N2 {
    class C {}
}

namespace N3 {
    class C {}
}

namespace N4 {
    using N1;
    using N2;
    
    class C3 {
        C Field; // either `N1.C` or `N2.C`, not `N3.C`
    }
}

namespace N5 {
    using N1;
    using N3;
    
    class C3 {
        C Field; // either `N1.C` or `N3.C`, not `N2.C`
    }
}
```

AFAICT, `N1.C`, `N2.C`, and `N3.C` will be considered equal modulo `getAnAmbiguousAlternativeType`, but that is an over approximation

Reply:
No, your understanding is correct, we over approximate. What would be your suggestion? (Let's continue this discussion on Slack)